From bbbd27a7939035de9ae65116561fd5de8039b81c Mon Sep 17 00:00:00 2001
Message-ID: <bbbd27a7939035de9ae65116561fd5de8039b81c.1779975966.git.bertrand.marquis@arm.com>
In-Reply-To: <55d56a174a0f04520d4a1558630bb0b739074b13.1779975966.git.bertrand.marquis@arm.com>
References: <55d56a174a0f04520d4a1558630bb0b739074b13.1779975966.git.bertrand.marquis@arm.com>
From: Bertrand Marquis <bertrand.marquis@arm.com>
Date: Fri, 22 May 2026 10:30:30 +0200
Subject: [PATCH 06/30] memory: Add direct-map-only address spaces

Add an AddressSpace policy bit for DMA users that require real direct
mappings, plus a setter for callers that need to opt in.

When the policy is enabled, address_space_map() rejects translations that
would allocate bounce buffers, and address_space_cache_init() rejects
indirect mappings while leaving the cache empty. The virtio-msg bridge DMA
address space uses this to preserve shared-memory semantics instead of
silently substituting a bounce buffer.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>
---
 include/system/memory.h | 19 +++++++++++++++++++
 system/memory.c         |  6 ++++++
 system/physmem.c        | 12 ++++++++++++
 3 files changed, 37 insertions(+)

diff --git a/include/system/memory.h b/include/system/memory.h
index 0562af313615..ebf15c566538 100644
--- a/include/system/memory.h
+++ b/include/system/memory.h
@@ -1179,6 +1179,13 @@ struct AddressSpace {
     QTAILQ_HEAD(, MemoryListener) listeners;
     QTAILQ_ENTRY(AddressSpace) address_spaces_link;
 
+    /*
+     * Address-space policy for DMA callers that require direct mappings.
+     * When enabled, address_space_map() and address_space_cache_init() reject
+     * mappings that would need an indirect bounce buffer because bounce buffers
+     * are semantically wrong for those callers.
+     */
+    bool direct_map_only;
     /*
      * Maximum DMA bounce buffer size used for indirect memory map requests.
      * This limits the total size of bounce buffer allocations made for
@@ -2704,6 +2711,18 @@ MemTxResult memory_region_dispatch_write(MemoryRegion *mr,
  */
 void address_space_init(AddressSpace *as, MemoryRegion *root, const char *name);
 
+/**
+ * address_space_set_direct_map_only: reject bounce buffers for an address space
+ *
+ * @as: address space to configure
+ * @direct_only: true to reject mappings that require bounce buffers
+ *
+ * Configure whether address_space_map() and address_space_cache_init() must
+ * return failure instead of creating indirect bounce-buffer mappings.  This is
+ * intended for DMA users where bounce buffers are semantically wrong.
+ */
+void address_space_set_direct_map_only(AddressSpace *as, bool direct_only);
+
 /**
  * address_space_destroy: destroy an address space
  *
diff --git a/system/memory.c b/system/memory.c
index c51d0798a843..7fee5a19d5ce 100644
--- a/system/memory.c
+++ b/system/memory.c
@@ -3231,6 +3231,7 @@ void address_space_init(AddressSpace *as, MemoryRegion *root, const char *name)
     as->ioeventfds = NULL;
     QTAILQ_INIT(&as->listeners);
     QTAILQ_INSERT_TAIL(&address_spaces, as, address_spaces_link);
+    as->direct_map_only = false;
     as->max_bounce_buffer_size = DEFAULT_MAX_BOUNCE_BUFFER_SIZE;
     as->bounce_buffer_size = 0;
     qemu_mutex_init(&as->map_client_list_lock);
@@ -3240,6 +3241,11 @@ void address_space_init(AddressSpace *as, MemoryRegion *root, const char *name)
     address_space_update_ioeventfds(as);
 }
 
+void address_space_set_direct_map_only(AddressSpace *as, bool direct_only)
+{
+    as->direct_map_only = direct_only;
+}
+
 static void do_address_space_destroy(AddressSpace *as)
 {
     assert(qatomic_read(&as->bounce_buffer_size) == 0);
diff --git a/system/physmem.c b/system/physmem.c
index 3a6d01e52fe3..132c662f6b42 100644
--- a/system/physmem.c
+++ b/system/physmem.c
@@ -3731,6 +3731,11 @@ void *address_space_map(AddressSpace *as,
     mr = flatview_translate(fv, addr, &xlat, &l, is_write, attrs);
 
     if (!memory_access_is_direct(mr, is_write, attrs)) {
+        if (as->direct_map_only) {
+            *plen = 0;
+            return NULL;
+        }
+
         size_t used = qatomic_read(&as->bounce_buffer_size);
         for (;;) {
             hwaddr alloc = MIN(as->max_bounce_buffer_size - used, l);
@@ -3846,6 +3851,7 @@ int64_t address_space_cache_init(MemoryRegionCache *cache,
     Int128 diff;
 
     assert(len > 0);
+    address_space_cache_init_empty(cache);
 
     l = len;
     cache->fv = address_space_get_flatview(as);
@@ -3874,6 +3880,12 @@ int64_t address_space_cache_init(MemoryRegionCache *cache,
         cache->ptr = qemu_ram_ptr_length(mr->ram_block, cache->xlat, &l, true,
                                          is_write);
     } else {
+        if (as->direct_map_only) {
+            memory_region_unref(mr);
+            flatview_unref(cache->fv);
+            address_space_cache_init_empty(cache);
+            return 0;
+        }
         cache->ptr = NULL;
     }
 
-- 
2.53.0

